Leaving a group or guild no longer removes you from challenges associated with those groups (merged in #8517)

Hus274 opened the pull request:
_edit by Alys 2017-02-28: This was accepted and was merged in https://github.com/HabitRPG/habitica/pull/8517_

closes #7764
### Changes

Leaving a group or a guild no longer removes the user from the challenges of that group or guild.

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

crookedneighbor commented on website/server/models/challenge.js:
```diff
@@ -240,13 +240,14 @@ schema.methods.unlinkTasks = async function challengeUnlinkTasks (user, keep) {
   };
 
   removeFromArray(user.challenges, challengeId);
+  this.memberCount--;
```
Good catch.

crookedneighbor commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -329,7 +329,6 @@ api.rejectGroupInvite = {
  * @apiGroup Group
  *
  * @apiParam {string} groupId The group _id ('party' for the user party and 'habitrpg' for tavern are accepted)
- * @apiParam {string="remove-all","keep-all"} keep Query parameter - Whether to keep or not challenges' tasks. Defaults to keep-all
```
Removing this behavior is a breaking change. So we can't do it. I think it would be better to correct this behavior and have `keep-all` keep challenges _and_ tasks, while `remove-all` should remove challenges _and_tasks. Does that make sense?

Hus274 replied:
Yeah that makes sense, I can look into that.

crookedneighbor commented:
@Hus274 What's the status on this?

Hus274 commented:
In progress. Going to try and get it finished up this week.

Commit:
Leaving a group or a guild no longer removes the user from the challenges of that group or guild.

Hus274 commented:
I've updated this to not be a breaking change and migrated to client-old.

crookedneighbor commented on website/server/models/group.js:
```diff
@@ -750,15 +750,18 @@ schema.methods.leave = async function leaveGroup (user, keep = 'keep-all') {
-    group: group._id,
-  });
+  // only remove user from challenges if it's set to remove-all
+  if (keep === 'remove-all') {
```
I think the API docs for the group leave route need to be updated https://github.com/HabitRPG/habitrpg/blob/b54441a637b91ddd80e370cd78885514c8056b3e/website/server/controllers/api-v3/groups.js#L362-L372

Since `keep` no longer defaults to anything, right? We should also clarify better what the keep queryParam means when passing it in.

Hus274 replied:
I've updated the documentation here. I left the "defaults to keep-all" section because that's still true: https://github.com/HabitRPG/habitrpg/pull/7781/files#diff-b6f239870392ed0ec2a2b73cbebaa699R747

crookedneighbor replied:
Hmm... See, in prod now, you always leave the challenge and the `keep` query param determines if you keep the tasks or have them removed. I _think_ the change you're doing is technically a breaking change.

I think it makes _more_ sense to provide a body parameter for keeping the challenges (in which case the query param is ignored). Then we just update the web app's calls to provide that parameter. 

3rd parties using this route will still remove all the challenges, but it won't be a breaking change.

crookedneighbor reviewed (changes requested):
Looks good, other than my one comment about documentation. :+1:

Commit:
Merge remote-tracking branch 'upstream/develop' into 7764

Commit:
Updating api docs for leaving group to take into account the default path no longer leaving challenges when leaving a group.

Commit:
Updating api docs for leaving group to take into account the default path no longer leaving challenges when leaving a group.

Hus274 commented:
I've updated the documentation, let me know if this works.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_leave.js:
```diff
@@ -93,7 +93,7 @@ describe('POST /groups/:groupId/leave', () => {
           let userWithChallengeTasks = await leader.get('/user');
 
-          expect(userWithChallengeTasks.challenges).to.not.include(challenge._id);
+          expect(userWithChallengeTasks.challenges).to.include(challenge._id);
```
I think the it block of this test should specify that the user remains in the challenge.

crookedneighbor commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -366,7 +366,7 @@ api.rejectGroupInvite = {
  *
  * @apiParam {String} groupId The group _id ('party' for the user party and 'habitrpg' for tavern are accepted)
- * @apiParam {String="remove-all","keep-all"} keep Query parameter - Whether to keep or not challenges' tasks. Defaults to keep-all
+ * @apiParam {String="remove-all","keep-all"} keep Query parameter - Whether or not to keep challenges belonging to the group being left. Defaults to keep-all
```
This should follow the APIDoc way of designating a default param: http://apidocjs.com/#param-api-param and use the Query param group (newly created)

``` js
* @apiParam (Query) {String="remove-all","keep-all"} keep=keep-all  Whether or not to keep challenges belonging to the group being left.
```

khipkin commented:
@Hus274 What's the status of this?

Hus274 commented:
Ugh, I definitely lost track of this one. I might be able to look into this this week but most likely won't be able to until next week as work is crazy now currently. If you'd like to pick it up, feel free as it looks like there's a fair bit of refactoring to do. Otherwise I'll take a look when I get a chance.

khipkin commented:
@Hus274 Nah it's totally yours! I just wanted to check that you hadn't forgotten about it, since this is a PR that I def already thought was merged into dev. 😊

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 7764

Hus274 committed:
refactored according to blade's comments to not be a breaking change. The api now accepts a body parameter to specify wether the user
should remain in the groups challenges or leave them. The change also adds more tests around this behavior to confirm that it works
as expected.

Hus274 commented:
I've updated this according to @crookedneighbor 's comments so it's not a breaking change by using a body parameter that the client side now sends that keeps the user in the group's challenges. I've also updated the test coverage to be slightly more specific and to cover the new case.

Let me know if anything else is needed here!

paglias commented on website/server/models/group.js:
```diff
@@ -833,21 +833,24 @@ schema.statics.tavernBoss = async function tavernBoss (user, progress) {
+    let challenges = await Challenge.find({
+      _id: {$in: user.challenges},
+      group: group._id,
+    });
```
missing .exec() (Note: I'm commenting on all PRs that have missing `.exec()` calls, not reviewing entire PRs)

Hus274 committed:
Merge remote-tracking branch 'upstream/develop' into 7764

Hus274 committed:
Merge branch 'develop' into 7764

Hus274 commented:
Fixed the merge conflicts here and added the exec() method to the end of the mongoose call.

khipkin commented:
Ping to reviewers!

TheHollidayInn commented:
Noted towards your Tier!

Alys commented:
FYI for the record:

We're having some trouble on our test server with the PRs that have been merged in recently, so I've started testing each one individually.  I tested this one on my local install by fetching upstream, then checking out 68a042cdb97bf00b496328a79660be97d150b660 and doing `npm start`. My testing has been VERY brief because I just wanted to follow on from my comments at https://github.com/HabitRPG/habitica/pull/8469#issuecomment-283013476

Testing locally with 68a042cdb97bf00b496328a79660be97d150b660 shows that tasks CAN be edited so the edit bug described in https://github.com/HabitRPG/habitica/pull/8469#issuecomment-283013476 does not come from this PR.

Alys commented:
I've had two reports that this bug is still happening so I'm reopening this. I'll test it myself later when I get a chance.

Alys commented:
Ugh, sorry, that was meant to go on the relevant issue.
